### ISSUE TITLE
Set up release Trilinos Windows nightly build

### ISSUE THREAD

**Issue by jwillenbring** [trilinos/Trilinos#1160]
@trilinos/framework 

This ticket is for setting up a nightly "release" build for Trilinos on Windows.

Done for this story includes

-having a nightly build reporting to CDash with a release configuration (with package options previously sent to Michael).

-the build should run on a Windows VM provided by the Trilinos team

-filing tickets for all errors found in executing the build. For all errors found, tickets should be filed in GitHub issues. If cause of error is known, a pull request could be offered to the package developers for package encountering the failure.

If necessary, another ticket will be filed for cleaning up all failures once the job is set up and running.

The initial work for setting up the build can be done anywhere, but getting the job running through Jenkins on the appropriate VM depends on tickets #1157 and #1158.

**Comment by mplooster** [comment]
@bmpersc @jwillenbring  @william76 What do you think about using the [ninja build system](https://ninja-build.org/) to parallelize the build on Windows? It works with MSVC and certainly speeds up compile time.

**Comment by bartlettroscoe** [comment]
> What do you think about using the ninja build system to parallelize the build on Windows?

That is the direction that we should go.  Work is being done to make Trilinos (and every TriBITS-based project) to be able to seamlessly use Ninja.  See:

* https://github.com/TriBITSPub/TriBITS/issues/72
* https://github.com/TriBITSPub/TriBITS/issues/166
* https://github.com/TriBITSPub/TriBITS/issues/167

After all of that is done, it will be easy to enable Ninja and most people may not even notice it is being used.

**Comment by mhoemmen** [comment]
I would like this, as long as it's easy to run the Windows VM on Linux or Mac so that we can replicate errors.

**Comment by mplooster** [comment]
Depends on Issue #1178. Submitted pull request #1179 as a potential fix.

**Comment by mplooster** [comment]
@trilinos/framework @trilinos/zoltan Is Perl required to build testing for Zoltan? I saw this in my configure output and now am wondering if I will need to include Perl in the Windows build.

> Processing enabled package: Zoltan (Libs, Tests)
-- Skipping Zoltan tests because PERL_EXECUTABLE is not found.

**Comment by kddevin** [comment]
Yes, the zoltan regression tests require perl.

**Comment by mplooster** [comment]
@trilinos/framework When I configure with Pamgen turned on, I get the message

> WARNING: User has set Trilinos_ENABLE_Pamgen=ON but the package Pamgen is not supported on this platform type 'Windows'!

I don't get any build errors, so I am wondering, why is Pamgen marked as unsupported for Windows?

**Comment by kddevin** [comment]
@mplooster did you decide to include perl in the Zoltan testing?  I got the impression from #1205 that you had excluded the tests that needed perl.  If so, how did you make CMake exclude those tests?  Thanks.

**Comment by mplooster** [comment]
@kddevin 

> did you decide to include perl in the Zoltan testing?

I downloaded Perl and told CMake where it was. Whether it was actually used for the Zoltan testing is unclear to me at this point.

> I got the impression from #1205 that you had excluded the tests that needed perl. If so, how did you make CMake exclude those tests?

I am building with MPI turned on, so I assume many of those tests were turned off because they were marked as serial. My configuration output seems to indicate as much (@trilinos/framework I will note that this is what it looks like for the Pamgen tests, too). Here is a sample:

> ...
-- Zoltan_ch_brack2_3_zoltan_serial: NOT added test because TPL_ENABLE_MPI='ON' and COMM='serial'!
-- Zoltan_ch_bug_zoltan_serial: NOT added test because TPL_ENABLE_MPI='ON' and COMM='serial'!
-- Zoltan_ch_degenerate_zoltan_serial: NOT added test because TPL_ENABLE_MPI='ON' and COMM='serial'!
-- Zoltan_ch_degenerateAA_zoltan_serial: NOT added test because TPL_ENABLE_MPI='ON' and COMM='serial'!
-- Zoltan_ch_drake_zoltan_serial: NOT added test because TPL_ENABLE_MPI='ON' and COMM='serial'!
...

Is Perl not required when building with MPI?

**Comment by kddevin** [comment]
If tests that have "parallel" in the name ran when MPI was enabled, then you used your perl installation.  E.g.,  Zoltan_ch_drake_zoltan_parallel

perl is needed for almost all zoltan tests, serial or parallel.  There are just a few that don't need it.

Thanks for your reply.  I was interested in knowing whether or not you had chosen to exclude all the tests that require perl.  It sounds as if you decided to keep them.  That is the best result.

**Comment by bmpersc** [comment]
@mplooster the files that we used to run testing on our previous windows machine are in Trilinos/cmake/ctest/drivers/kallikrates. The important files to look at would be TrilinosCTestDriverCore.Kallikrates.msvc.cmake and ctest_windows_nightly_serial_development.cmake. There are a number of other files in the directory that we were using to test with, but for the most part they were for a shared library build or other random tests we were or trying to run at the time.

ctest_windows_nightly_serial_development.cmake contains a lot of the specific configure options and ctest settings for the specific build. This is the script that we actually call to run the test with ctest -S.

TrilinosCTestDriverCore.Kallikrates.msvc.cmake contains more generic things for the system like the location of TPLs, compiler information, etc. It is called by the ctest_windows_nightly_serial_development.cmake script, but it actually controls a lot of the actual test run work.

I would not recommend modifying these files directly since you are on a machine very different than what kallikrates was, but they could be a good start to a new set of generic-ish files. For the short term you might want to consider making a new directory in the cmake/ctest/drivers directory for windows and put your scripts there. The scripts are, unfortunately, written in a way that does expect things to be in a specific directory relative to the current so the directory depth is important.

**Comment by mplooster** [comment]
To use the TriBITS ctest driver mechansim, TriBITSPub/TriBITS#195 needs to be fixed. Otherwise, the optional packages will always be built (and fail if they are not supported on Windows).

**Comment by bartlettroscoe** [comment]
I will take a look at TriBITSPub/TriBITS#195 shortly.  I will update it once I fix it.

**Comment by bartlettroscoe** [comment]
@mplooster,

Okay, I can see that `${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES=ON` is being hard-coded and perhaps we should allow the user to change this.  But I have to ask why this is needed?  If you disable `${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES` then I don't see how the package-by-package testing process will correctly test each package.  It would only be testing a given package against just its required packages which is not enough testing (even for Windows).  Based on the packages you want test test, we can look at what would happen if you set  `${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES=OFF`.

Instead, can you consider using blacklisting?  That is where you leave `${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES` turned on and then specifically disable the packages that you don't want to test.   You can see an example of this, for instance, in commit  b47e6ad4f1. That will behave correctly in package-by-package testing.  If it is not clear why that should be a much better approach, then let's talk.  (You can contact me offline, send email rabartl at sandia dot gov and we can set it up.)

**Comment by mplooster** [comment]
@bartlettroscoe 

> It would only be testing a given package against just its required packages which is not enough testing (even for Windows).

Perhaps @jwillenbring can comment here, but for this task, there is only a specific set of packages that we want to build, so yes, this is the desired behavior in this case. I believe the underlying reason is that we are not yet at a place to make all the changes that would allow the optional packages to build _and test_ correctly on Windows.

I will also point out that there are existing ctest driver scripts in Trilinos that attempt to turn this off, which tells me there is a need (and/or desire) for this functionality to work. For example,

[ctest_linux_nightly_serial_release_core_all_kokkos-dev.cmake](https://github.com/trilinos/Trilinos/blob/f0350316239aaf41e8e6b82378612aaedf9cc72c/cmake/ctest/drivers/kokkos-dev/ctest_linux_nightly_serial_release_core_all_kokkos-dev.cmake#L98)

[ctest_linux_nightly_serial_performance_godel.cmake](https://github.com/trilinos/Trilinos/blob/f0350316239aaf41e8e6b82378612aaedf9cc72c/cmake/ctest/drivers/godel/ctest_linux_nightly_serial_performance_godel.cmake#L77)

[ctest_linux_nightly_mpi_development_opt_gcc_4.4.7_rhel6-x86_64_xyce.cmake](https://github.com/trilinos/Trilinos/blob/8f5c62fa53b268d0a104e57e52073dc1347a3426/cmake/ctest/drivers/rhel6-x86_64/MPI_OPT_GCC_4.4.7/ctest_linux_nightly_mpi_development_opt_gcc_4.4.7_rhel6-x86_64_xyce.cmake#L79)

[ctest_linux_nightly_generic_sierra.cmake](https://github.com/trilinos/Trilinos/blob/10547a3f7628d055d0ecee56e9d16807ae0d74f6/cmake/ctest/drivers/parameterized/ctest_linux_nightly_generic_sierra.cmake#L76)

[ctest_windows_nightly_serial_development_shared.cmake](https://github.com/trilinos/Trilinos/blob/f0350316239aaf41e8e6b82378612aaedf9cc72c/cmake/ctest/drivers/kallikrates/ctest_windows_nightly_serial_development_shared.cmake#L78)

[ctest_windows_nightly_serial_release_shared.cmake](https://github.com/trilinos/Trilinos/blob/f0350316239aaf41e8e6b82378612aaedf9cc72c/cmake/ctest/drivers/kallikrates/ctest_windows_nightly_serial_release_shared.cmake#L81)

> Instead, can you consider using blacklisting? That is where you leave ${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES turned on and then specifically disable the packages that you don't want to test.

We could, but it would likely be a large number of packages, and the end result would probably be the same as if the ${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES variable worked for the ctest driver.

**Comment by bartlettroscoe** [comment]
> there is only a specific set of packages that we want to build, so yes, this is the desired behavior in this case. I believe the underlying reason is that we are not yet at a place to make all the changes that would allow the optional packages to build and test correctly on Windows.

Specifically, what is the exact set of packages you want to test?   We need to make this conversation concrete.

> I will also point out that there are existing ctest driver scripts in Trilinos that attempt to turn this off, which tells me there is a need (and/or desire) for this functionality to work. 

No, it just means whoever wrote those did not realize that their setting this has no effect and they did not understand how packages-by-package testing works.

> We could, but it would likely be a large number of packages, and the end result would probably be the same as if the `${PROJECT_NAME}_ENABLE_ALL_OPTIONAL_PACKAGES` variable worked for the ctest driver.

What you do is you specifically list the packages you want to test in the CTest var `Trilinos_PACKAGES` and then any packages you don't want to be tested that might get enabled due to `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES`, you disable them.  You can see a bunch of examples of this by looking at:

```
$ find cmake/ctest/drivers/ -name "*.cmake" -exec grep -nH "\(Trilinos_PACKAGES\|Trilinos_EXCLUDE_PACKAGES\)" {} \; 
```

Tell me the exact list of packages you want to test and don't want to test that get enabled as optional upstream packages and I will give you the CTest code to add.  It will be very little code.

**Comment by mplooster** [comment]
> Specifically, what is the exact set of packages you want to test? We need to make this conversation concrete.

See this attached example script: [CTestConfigureEpetraStack.txt](https://github.com/trilinos/Trilinos/files/1042596/CTestConfigureEpetraStack.txt).

You will notice that I am already using the Trilinos_EXCLUDE_PACKAGES and -DTrilinos_ENABLE_\<Package\> options to blacklist a couple of packages.

I understand the point that disabling all optional packages doesn't give as much test coverage. However, for this particular build that we are setting up at this time, we really do only want the set of listed packages, nothing more or less. I can certainly blacklist the dozen or so optional packages that are enabled (which is fine, it's a valid way to do it), but it would be more convenient if I could just turn off the optional packages and only enable the ones I explicitly want.

**Comment by bartlettroscoe** [comment]
Thanks for making this concrete.  If you disable all optional packages, then when Belos, for example, is processed by TribitsCTestDriverCore.cmake, it will only enable Teuchos and not any of the solvers that depend on Epetra, for instance.  You can see this from the file packages/belos/cmake/Dependencies.cmake.  That is **not** what you want.

If you don't believe me, go ahead and modify your local copy of TriBITS on your branch under Trilinos/cmake/tribits/ and do a trial experimental submit and see what happens on CDash.  I think after you do that experiment hopefully this point should be made clear.

Blacklisting the packages you don't want to test should only take 5 minutes.

**Comment by bmpersc** [comment]
The problem with depending on a blacklist is that what needs to be blacklisted will change as packages' dependencies change. So for right now it might only be 5 minutes, but we will pay that cost over and over again. The way we will discover that a new package will need to be added is that the build will fail with seemingly no relevant changes having been made. What is worse is if a dependency is added that doesn't fail the build we might not notice and the new dependency could change behavior that is unacceptable to the customer in question.

With a whitelist we'd be able to set exactly what we want once and not have to worry about the development that is tangential to the needs of the build.

**Comment by bartlettroscoe** [comment]
> The problem with depending on a blacklist is that what needs to be blacklisted will change as packages' dependencies change. So for right now it might only be 5 minutes, but we will pay that cost over and over again. The way we will discover that a new package will need to be added is that the build will fail with seemingly no relevant changes having been made. What is worse is if a dependency is added that doesn't fail the build we might not notice and the new dependency could change behavior that is unacceptable to the customer in question.
>
> With a whitelist we'd be able to set exactly what we want once and not have to worry about the development that is tangential to the needs of the build.

I disagree.  It is much better to be enabling and testing too much and have to disable things you don't want than to be enabling and testing too little (which will be hard to notice).  The larger and more complex the package dependency graph gets, the more blacklisting shines over white-listing.  CASL VERA has shown that out for many years now.

**Comment by mplooster** [comment]
Okay, I sense a lot of pushback on this. I will use blacklisting to achieve the same thing. It's not difficult, and I personally don't have a problem with doing it. However, for the sake of understanding I have to ask, if testing the **optional** packages is so important, why are they **optional** in the first place (instead of required)?

**Comment by bartlettroscoe** [comment]
> Okay, I sense a lot of pushback on this. I will use blacklisting to achieve the same thing. It's not difficult, and I personally don't have a problem with doing it.

As I said, we can implement what is being asked for here.  It is just that I can guarantee that it will not have the behavior you expect.   For example, do you really want to be testing Belos without Epetra?   I have not done a good job of making this point so now I can try to drive this home as follows.  If you turn off optional packages in the inner configuration of Trilinos with the TribitsCTestDriverCore.cmake code, it will configure, build, and test Belos, for example, as follows ...

Using the standard Trilinos CI build, using the `do-configure` script:

```
#!/bin/bash

cmake \
-DTPL_ENABLE_MPI=ON \
-DCMAKE_BUILD_TYPE=RELEASE \
-DTrilinos_ENABLE_DEBUG=ON \
-DBUILD_SHARED_LIBS=ON \
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE=OFF \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake \
-DTrilinos_ENABLE_DEBUG_SYMBOLS=ON \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-DTrilinos_ENABLE_CONFIGURE_TIMING=ON \
-DTrilinos_ENABLE_TESTS=ON \
-DTrilinos_TRACE_ADD_TEST=ON \
"$@" \
../../../Trilinos

```

the configure of Belos would be:

````
$ ./do-configure -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
  -DTrilinos_ENABLE_Belos=ON &> configure.out
````

This results in the enables (shown in configure.out):

```
Final set of enabled packages:  Teuchos Belos 2

Final set of enabled SE packages:  TeuchosCore TeuchosParameterList TeuchosComm TeuchosNumerics TeuchosRemainder Teuchos Belos 7
```

After building with `make -j16` the tests run for Belos then are:

```
$  ctest
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.9.3/MPI_RELEASE_DEBUG_SHARED_PT
    Start 1: Belos_Factory_MPI_4
1/1 Test #1: Belos_Factory_MPI_4 ..............   Passed    1.08 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
Belos    =   1.08 sec (1 test)
```

See, you are really not testing anything in Belos! 

Does this make more sense now?

I think what is unclear here is the workings of the package-by-package configures, builds, and testing that TribitsCTestCoreDriver.cmake does.  That is what I am trying to explain.

> However, for the sake of understanding I have to ask, if testing the optional packages is so important, why are they optional in the first place (instead of required)?

Having dependencies be optional gives the the flexibility to build only what you want.  For your testing, you want to test Belos with Epetra support, but not with Tpetra.  But Albany, for example, wants to test Belos with Tpetra and not Epetra.  The only way to support both of those use cases is for Belos to have optional dependencies on Epetra *and* Tpetra.  But if you test Belos with only its required dependencies, you get nothing.  Does that make sense?

**Comment by mplooster** [comment]
Thanks for the explanation. Yes, I see what you are pointing out. I think it is unfortunate that in the case of Belos, the optional packages are really not optional for any meaningful testing to happen. Because I must have Epetra or Tpetra for Belos, what I really have, IMO, is a choice of required packages.

So my next question is, why doesn't the ctest driver take it into account when building Belos that I have also specified Epetra in Trilinos_PACKAGES? Or for that matter, could I not enable Epetra for all the packages by adding it as a configure option, which would solve the problem for Belos?

```
SET(EXTRA_SYSTEM_CONFIGURE_OPTIONS
"-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF"
"-DTrilinos_ENABLE_Epetra:BOOL=ON"
...

```

**Comment by bartlettroscoe** [comment]
>  I think it is unfortunate that in the case of Belos, the optional packages are really not optional for any meaningful testing to happen. Because I must have Epetra or Tpetra for Belos, what I really have, IMO, is a choice of required packages.

That is really not true.  There are customers of Belos that don't use Epetra or Tpetra and instead instantiate with their own types.  Therefore, Belos can be useful for a customer even if they don't use Epetra or Tpetra.  Therefore, they really are optional dependencies.

> So my next question is, why doesn't the ctest driver take it into account when building Belos that I have also specified Epetra in Trilinos_PACKAGES? 

That would require a bit more logic and, yes, more complexity.  I have been told many times that there is too much complexity in TribitsCTestDriverCore.cmake (which is why I am answering these questions and not other Trilinos framework team members).

But once we implement TriBITSPub/TriBITS#183, you will have the flexibility to do white-listing (independent if that is better than blacklisting or not).

**Comment by bartlettroscoe** [comment]
@mplooster, did you resolve this?

**Comment by mplooster** [comment]
@bartlettroscoe  I went ahead with the blacklisting approach for now.

**Comment by bmpersc** [comment]
I have set the test that @mplooster had created to be run automatically through our jenkins server. It is currently set to run nightly and report to the experimental track.

There were a couple issues I had to address to get them to work with jenkins, but they were relatively minor, but one was a bit obscure. First the URL for the Trilinos repo was being set to the ssh version which required authentication tokens with github that I didn't want to set up for these builds since that shouldn't be necessary. I switched the URL back to the default https. This led to the obscure issue. The clone done by tribits wouldn't work with no reason given. It turns out the clone was timing out due to not being able to connect because of the proxies. I had to set the proxy information directly in the git configure, but once I did that it seemed to work and the rest of the builds worked mostly how I expected.

There are still some outstanding build and test errors, but we were expecting that as making every test clean was not part of the requirements. I am leaving the builds in the experimental track until we've either addressed the errors or have a plan that will get them addressed. At that point the builds will probably be moved to specialized, but the exact track can be determined at that time.

The build errors that are still outstanding look to be related to a limitation in windows. The error is:
fatal error C1083: Cannot open compiler generated file: '': Invalid argument
which after a bit of googling seems like it is probably related to the length of the filename being beyond the max filename length windows supports. This likely has a lot to do with the directory structure and will need to be investigated to see what can be done. I ran into this issue while installing compilers and tools so I'm not real surprised this happened. We will track that issue in a separate ticket since it isn't directly related to getting the testing setup.
